synapse/handlers/saml_handler.py

```diff
@@ -174,26 +177,56 @@ def handle_redirect_request(self, client_redirect_url):
                     return registered_user_id
 
             # figure out a new mxid for this user
-            base_mxid_localpart = self._mxid_mapper(mxid_source)
+            for i in range(1000):
+                # Use the provider's custom handler if available
+                if self._mapping_provider:
+                    localpart = self._mapping_provider.mxid_source_to_mixd_localpart(
+                        mxid_source, i
+                    )
+                else:
+                    localpart = self.mxid_source_to_mxid_localpart(mxid_source, i)
+                logger.info("Allocating mxid for new user with localpart %s", localpart)
 
-            suffix = 0
-            while True:
-                localpart = base_mxid_localpart + (str(suffix) if suffix else "")
+                # Check if this mxid already exists
                 if not await self._datastore.get_users_by_id_case_insensitive(
                     UserID(localpart, self._hostname).to_string()
                 ):
+                    # This mxid is free
                     break
-                suffix += 1
-            logger.info("Allocating mxid for new user with localpart %s", localpart)
+            else:
+                # Unable to generate a username in 1000 iterations
+                # Break and return error to the user
+                raise SynapseError(
+                    500, "Unable to generate a Matrix ID from the SAML response"
+                )
 
             registered_user_id = await self._registration_handler.register_user(
                 localpart=localpart, default_display_name=displayName
             )
+
             await self._datastore.record_user_external_id(
                 self._auth_provider_id, remote_user_id, registered_user_id
             )
             return registered_user_id
 
+    def mxid_source_to_mxid_localpart(self, mxid_source: str, failures: int = 0) -> str:
```

Comment:
it'd be nice to factor out a 'default_mxid_mapper' so that this is just an implementation of the module api. It'll simplify the code above and help make sure that the module API keeps working.

Reply:
So create a default provider class that we load in the config if one isn't specified? That sounds good.